cmd/snap-bootstrap/cmd_scan_disk.go

```diff
@@ -0,0 +1,290 @@
+		label := partition.GetName()
+		uuid := partition.GetUUID()
+		fmt.Fprintf(os.Stderr, "Found partition %s %s\n", label, uuid)
+		ret = append(ret, Partition{label, uuid})
+	}
+
+	return ret, nil
+}
+
+func samePath(a, b string) (bool, error) {
+	aSt, err := os.Stat(a)
+	if err != nil {
+		return false, err
+	}
+	bSt, err := os.Stat(b)
+	if err != nil {
+		return false, err
+	}
+	return os.SameFile(aSt, bSt), nil
+}
+
+func scanDiskNode(output io.Writer, node string) error {
+	/*
+	 * We need to find out if the give contains the ESP that was booted.
+	 * The boot loader will set LoaderDevicePartUUID. We will need to scan all
+	 * the partitions for that UUID.
+	 */
+	fmt.Fprintf(os.Stderr, "Scanning disk %s\n", node)
+	fallback := false
+	var fallbackPartition string
+
+	bootUUID, err := bootFindPartitionUUIDForBootedKernelDisk()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "No efi var, falling back: %s\n", err)
+		fallback = true
+	}
+
+	partitions, err := probePartitions(node)
+	if err != nil {
+		return fmt.Errorf("Cannot get partitions: %s\n", err)
+	}
+
+	/*
+	 * If LoaderDevicePartUUID was not set, it is probably because
+	 * we did not boot with UEFI. In that case we try to detect
+	 * disk with partition labels.
+	 */
+	if fallback {
+		mode, _, err := boot.ModeAndRecoverySystemFromKernelCommandLine()
+		if err != nil {
+			return err
+		}
+		switch mode {
+		case "recover":
+			fallbackPartition = "ubuntu-seed"
+		case "install":
+			fallbackPartition = "ubuntu-seed"
+		case "factory-reset":
+			fallbackPartition = "ubuntu-seed"
+		case "run":
+			fallbackPartition = "ubuntu-boot"
+		case "cloudimg-rootfs":
+			fallbackPartition = "ubuntu-boot"
+		default:
+			return fmt.Errorf("internal error: mode not handled")
+		}
+	}
+
+	/*
+	 * If we are not in UEFI mode and snapd_system_disk is
+	 * defined, we need to verify the disk also matches that. If
+	 * not, we just return, ignoring this disk.
+	 */
+	if fallback {
+		values, err := kcmdline.KeyValues("snapd_system_disk")
+		if err != nil {
+			return fmt.Errorf("Cannot read kernel command line: %s\n", err)
+		}
+
+		if value, ok := values["snapd_system_disk"]; ok {
+			var currentPath string
+			var expectedPath string
+			if strings.HasPrefix(value, "/dev/") || !strings.HasPrefix(value, "/") {
+				name := strings.TrimPrefix(value, "/dev/")
+				expectedPath = fmt.Sprintf("/dev/%s", name)
+				currentPath = node
+			} else {
+				expectedPath = value
+				currentPath = osGetenv("DEVPATH")
+			}
+
+			same, err := samePath(filepath.Join(dirs.GlobalRootDir, expectedPath),
+				filepath.Join(dirs.GlobalRootDir, currentPath))
+			if err != nil {
+				return fmt.Errorf("Cannot check snapd_system_disk kernel parameter: %s\n", err)
+			}
+			if !same {
+				return nil
+			}
+		}
+	}
+
+	/*
+	 * Now we scan the partitions. For each partition we need to find out:
+	 *  - If its label matches a known partition type. In this
+	 *    case we save the UUID of that partition.
```

Comment:
But we are not really saving the UUID, is this comment right?

Reply:
Not anymore. I have to tweak this.